### PR TITLE
Use strcmp instead of < in PHP: Add test for Reflect.compare and BalancedTree with string keys.

### PR DESCRIPTION
**Pull request by georgkoester**
Add test for Reflect.compare and BalancedTree with string keys.

### CODE REVIEW

**Comment by georgkoester** [comment]
I like the BalancedTree data structure because of its adaptive memory load (no growing-only hash array). Is that structure (haxe.ds.BalancedTree) part of the haxe data structures lib? It's not mentioned on haxe.org.
What about adding a Comparator parameter?

File: tests/unit/src/unitstd/haxe/ds/BalancedTree.unit.hx [hunk 1]
```diff
@@ -31,4 +31,26 @@ r == [4,8,2,5,1,6,3,9,7,10];
...
+	eq(test[k], ms.get(Std.string(k)));
+}
+for (k in test.keys()) {
+	eq(true, ms.exists(Std.string(k)));
+}
+for (k in otherKeys) {
+	eq(false, ms.exists(Std.string(k)));
+}
+var r = [for (key in ms.keys()) key];
+r == [for (k in [1,11,13,15,17,22,25,27,6,8]) Std.string(k)];
```
Comment (Simn) [inline tests/unit/src/unitstd/haxe/ds/BalancedTree.unit.hx]: Does testing like this actually work?

**Comment by Simn** [comment]
Passing comparator functions is quite slow on several targets. The intended usage is to subclass BalancedTree and override the compare method. This is also what EnumValueMap does.

**Comment by georgkoester** [comment]
In fact it should work - probably that todo came from Int64 not being comparable with Reflect.compare. To a newcomer it looks like a number type, too.

**Comment by georgkoester** [comment]
This now fails because in PHP the operators compare strings as integers if possible (even ignores preceeding whitespace). See [1] and try for yourself:
```
<?php

test("11", "3");
test("3", "2012");
test(" 1234", "1234");
test("\n1234", "1234");

function test($v1, $v2) {
    echo "strcmp of them: " . strcmp($v1, $v2) . "(neg value means " . $v1 . " less than " . $v2 . "\n";
    if($v1 < $v2) {
        echo $v1 . " less than " . $v2 . "\n";
    }
    else {
        echo $v1 . " not less than " . $v2 . "\n";
    }
}

function show_cr($var) {
    return str_replace("\n", "\\n", $var);
}

function my_var_dump($var) {
    ob_start();
    var_dump($var);
    $dump = show_cr(trim(ob_get_contents()));
    ob_end_clean();
    return $dump;
}

?>
```

**Comment by georgkoester** [comment]
If I get a direction I would look into a fix. Do I have to create a native string class or is it somewhere in the ocaml generation logic?

**Comment by Simn** [comment]
I don't really understand the issue. Why are we talking Strings if the failing tests involve Ints?

**Comment by georgkoester** [comment]
Hi @Simn , the failures are happening in the second part of BalancedTree.unit.hx where the BalancedTree uses String keys. The String keys are generated from the lists of integers used for the original test. So the strings are stringified ints. The test fails on php because the string keys are not ordered in the String-order ("2" is bigger than "11"), but in Int-order (2 is smaller than 11).

**Comment by Simn** [comment]
Ah okay, that's a bit confusing.

I suppose we have to fix this in Reflect.compare then.

**Comment by georgkoester** [comment]
Hi @Simn , I added another test to clarify that not only Reflect.compare needs fixing.

I also rebased.

And I created a docker image to simplify running unit tests (has some helpful scripts, sets up mysql for php and adds required packages). Also avoids creating too much trash on the developers machine: https://github.com/georgkoester/haxe_dev_dockerimage

**Comment by Simn** [comment]
That's interesting, I didn't even know we allowed comparison ops on String. Thanks for the work on this.

**Comment by Simn** [comment]
I've merged this locally and applied a fix. Thanks again!

**Comment by georgkoester** [comment]
Hi @Simn , thanks for your fix!

**Pull request closed without merging**